apache/pulsar
# [fix][client] Fix producer/consumer stop to reconnect or Pub/Sub due to IO thread race-condition

### Motivation

**Issue 1**: see the test `testUnstableNetWorkWhenCreatingProducer`

The producer fails to start if the network is unstable.

| time | socket closed | start producer |
| --- | --- | --- |
| 1 | | get a pooled connection from the pool |
| 2 | | before registering the producer to the connection |
| 3 | socket closed |
| 4 | reconnect all producers, which is empty now |
| 5 | | register the producer to a closed connection |
| 6 | | failed to send `Command.newProducer` to brokers |
| 7 | | users receive an error `io.netty.channel.StacklessClosedChannelException: null` |

---

<del>
**Issue 2**: see the test: `testSendAfterCreateProducerAsync`
- Start a producer/consumer async
- Call a sync method in the callback method of method `createAsync`, such as `producer.send`, and `consumer.receive`.
- The IO thread encounters a deadlock
</del>

### Modifications

- fix the issues above

### Documentation




### Matching PR in forked repository

PR in forked repository: x
